---
title: "2022_05_17_ShapeMaP_KimTD"
author: "Joon Yoon"
date: "`r Sys.Date()`"
html_document:
        code_folding: hide
        toc: yes
        toc_float: true
        toc_depth: 6
        number_sections: true
        theme: paper
---

```{r message=FALSE ,warning = FALSE,include=F}
# Shared R Markdown settings
if (file.exists("_setup.R")) {
    source("_setup.R")
}
```

```{r, message = FALSE, warning = FALSE}
# Avoid StartupMessages
shhh <- suppressPackageStartupMessages
```

```{r echo=FALSE, warning=FALSE, message=FALSE}
shhh(library("knitcitations"))
options("citation_format" = "pandoc")
clientname="Taedon Kim"
clientemail="tdkim@kribb.re.kr"
PIname=""
lablocation="Samsung Biologics"
analystname="Joon Yoon"
analystemail="joon17.yoon@samsung.com"

shhh(library(knitr))
opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, cache=TRUE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

---

Shapemapper and Superfold outputs for `r clientname` (`r clientemail`).  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

---

## Analysis pipeline
![](./image_files/pipeline_summary_shapemapper_kimtd.png)

```
For this report, please note that we evaluated the ShapeMaP outputs using Shapemapper2 and Superfold.
```

## 4-Q5 
### 4-Q5_Shapemapper Log
![](./image_files/4-q5-log.png)

```
This pair seems fine in terms of alignment rate and the shapemapper QC.
```

### 4-Q5_shapemapper Histograms
![](./image_files/kimtd_run2_Reference_KLF4_histograms.png)

```
The drop in the read depth seems concerning.
```

### 4-Q5_shapemapper Profiles
![](./image_files/kimtd_4Q-5_reference_KLF4_profiles.png)

```
We can clearly see that around position 120, there is a drop in read depth.
This has been confirmed through IGV of sam.
```

#### Confirmation with IGV
![](./image_files/IGV_screenshot.png)
```
When zoomed in, the drop happens where a repetitive base occurs.
```
#### Zoomed IGV
![](./image_files/4-Q5_drop_point.png)
```
This is where the drop occurs; 100~180bp position.
You could see the nucleotide based at the bottom.
```


### 4-Q5_shapemapper shannonShape plot
![](./image_files/shannonShape_kimtd_run2_Reference_KLF4_map_3ba9.png)

```
Black: SHAPE
RED: Shannon
Light Purple zone: low SHANNON low SHAPE

X-axis: Nucelotide position
left Y-axis: Shannon Entropy
right Y-axis: Shape Reactivity
```
### 4-Q5 Structure Comparison
![](./image_files/4-Q5_sequence_vs_Shapemap.png)

## 6-Pfu
![](./image_files/6-pfu-log.png)

```
QC did not pass for these samples. Please read the log for details.
```

### 6-Pfu_shapemapper Histograms
![](./image_files/kimtd_6pfu_Reference_KLF4_histograms.png)

```
These samples also display the depth drop.
```

### 6-Pfu_shapemapper Profiles
![](./image_files/kimtd_6pfu_Reference_KLF4_profiles.png)

```
We can clearly see that around position 110, there is a drop in read depth.
This has been confirmed through IGV of sam.
```

#### Zoomed IGV
![](./image_files/6-PFU_drop_point.png)
```
This is where the drop occurs; 100~180bp position.
You could see the nucleotide based at the bottom.
```

### 6-Pfu_shapemapper shannonShape plot
![](./image_files/shannonShape_kimtd_6pfu_Reference_KLF4_map_06ec.png)

```
Black: SHAPE
RED: Shannon
Light Purple zone: low SHANNON low SHAPE

X-axis: Nucelotide position
left Y-axis: Shannon Entropy
right Y-axis: Shape Reactivity
```

### 6-pfu Structure Comparison
![](./image_files/6-PFU_sequence_vs_Shapemap.png)

# Reference ReadME of Shapemapper
## Original Link

Please refer to [Shapemapper_github](https://github.com/Weeks-UNC/shapemapper2/blob/master/docs/analysis_steps.md) for full document.

## Partial ReadME
### Basecall quality score resolution

At a given position where mate pairs overlap, the higher basecall quality score
is used for the merged read.

### Mutation resolution
Mutations that overlap are first grouped. For each group of mutations,
the mean phred score over
the set of basecalls within and adjacent to R1 mutations
is compared to the mean phred score over R2 mutations
and adjacent positions. The read mutation group with the higher mean phred score is
then used for the merged read. This strategy avoids rare cases in which
a more naïve mutation resolution algorithm results in a merged read that is 
incompatible with both mate pairs.

### Ambiguously aligned mutation handling
Early versions of ShapeMapper entirely excluded ambiguously aligned mutations
from analysis, since they make it impossible to infer adduct location
with full single nucleotide accuracy. However, short ambiguous deletions
are a fairly large signal within many MaP datasets, and we find that
realigning ambiguous mutations 5′ and including them in analysis 
produces empirical improvements in reactivity profile accuracy. 
See SFig. 2 in [Busan and Weeks, 2018](http://rnajournal.cshlp.org/content/early/2017/11/07/rna.061945.117).
The default behavior (5′ realignment) can be reversed (3′ realignment) by
passing the 
<kbd>--right-align-ambig</kbd> 
option (not recommended).

*Example realignment:* A five-nucleotide deletion can be equivalently
aligned to two locations. ShapeMapper uses the 5′ location.

    3′ aligned read:  GAGGAAGGTGGGGATGACGTCA-----TCATGGCCCTTACG
    5′ aligned read:  GAGGAAGGTGGGGATGAC-----GTCATCATGGCCCTTACG
             target:  GAGGAAGGTGGGGATGACGTCAAGTCATCATGGCCCTTACG
                                            ^
                                            inferred adduct site

We hypothesize that most short ambiguous deletions are a result of 
reverse transcriptase encountering an adduct, then stalling until a 
few nucleotides of the cDNA strand locally melt and reanneal to another 
complementary position within the RNA a few nucleotides toward the 5′ end. 

Repeat and/or homopolymeric regions still pose difficulties for MaP, both due to
high background reverse transcriptase mutagenesis within these regions and 
due to frequent loss of information regarding adduct position. Future
ShapeMapper versions may track and warn the user about such regions.

### Multinucleotide mutation handling

We observe that  SuperScript II reverse transcription
under MaP conditions on SHAPE-modified RNA produces some reads
with multiple nearby mutations, and hypothesize that many of these
mutation clusters in fact reflect single adducts toward the 3′ end
of a given cluster. ShapeMapper therefore applies a mutation separation
threshold to collapse nearby mutations.

Mutations separated by less than <kbd>--min-mutation-separation</kbd> unchanged 
reference sequence nucleotides are collapsed and treated as a single mutation.
This is not applied within sequence correction stages if present (
<kbd>--correct-seq</kbd>
samples).
The default mutation separation threshold (6 nucleotides) was empirically 
selected to maximize reactivity profile accuracy for an 
extracted _E. coli_ rRNA SHAPE dataset
(see SFig. 1 in [Busan and Weeks, 2018](http://rnajournal.cshlp.org/content/early/2017/11/07/rna.061945.117)).
The same threshold also appears optimal for DMS adducts (unpublished data). Alternative 
reverse transcriptases, reverse transcription conditions, or adduct chemistries
may require re-optimization.

_Example:_ 

![multinuc mutation](images/multinuc_mutations2.png)

### Post-alignment basecall quality filtering

Initial read quality trimming is fairly permissive to support
high read depths and to permit reads with a few bad basecalls
to be used as long as bad basecalls are isolated (see
[Initial basecall quality trimming](#initial-basecall-quality-trimming)).

After alignment, ShapeMapper applies a separate, more stringent
basecall quality filter to remove mutations and read depths associated
with basecalls with quality scores below 
<kbd>--min-qual-to-count</kbd>. 
For sequencing runs of mixed quality (such as runs approaching the length
limits of Illumina chemistry), this strategy enables recovery of 
high read depths without sacrificing reactivity profile accuracy.
See SFig. 1 in [Busan and Weeks, 2018](http://rnajournal.cshlp.org/content/early/2017/11/07/rna.061945.117).

Specifically, 
for each nucleotide position not within mutations (that is, a position covered by a read
but unchanged from reference sequence), a given position is excluded if its basecall,
the basecall immediately 5′, or the basecall immediately 3′ has
a quality score below <kbd>--min-qual</kbd>.

A given mutation is excluded (and the covered sequence region) if it contains
any basecall with quality score below <kbd>--min-qual</kbd>, or if 
the basecall immediately 5′ or the basecall immediately 3′ has
a quality score below <kbd>--min-qual</kbd>.


_Example 1:_

![basecall qualities simple](images/basecall_quality1.png)

A poorly scoring portion of the read is excluded from analysis.

_Example 2:_

![basecall qualities example](images/basecall_quality2.png)

Poorly scoring regions of the read are excluded, and one mutation 
associated with low-quality basecalls is excluded.

#### Chemical adduct location inference

ShapeMapper infers the locations of chemical adducts as the reference 
sequence position immediately 5′ of the last unchanged reference
nucleotide before a mutation scanning 3′→5′ (as in reverse transcription).

_Examples:_

![adduct locations](images/adduct_inference1.png)

### Calculation of mutation rates

#### Mutation and read depth counting
After all mutation and read depth processing and filtering
steps are performed, ShapeMapper sums mutation counts and
read depths for each position in each RNA sequence.

#### Effective read depth
Simple sequencing read depths are computed as the number of reads crossing
a given position in the target sequence. For computing mutation rates,
however, these depths are somewhat inadequate. The mutation rate at a 
given position should be computed as the number of observed mutations
divided by the total number of observations (i.e. the number of
"no mutation" observations plus the number of "mutation" observations).

A given position within the region covered by a multinucleotide mutation is
not actually an observation of "no mutation" - we are effectively blind
over these regions in individual reads. To help account for this, the region over
each mutation is excluded from contributing to the effective read depth, 
with the exception of the inferred adduct site (see examples above). This
correction is not applied to sequence correction stages. In practice, 
excluding multinucleotide sites from read depths is usually a small change 
to a large denominator, and has a negligable effect on reactivity profile 
accuracy, at least with mutation rates in the range produced by SHAPE adducts. 

Effective read depths as reported by ShapeMapper also exclude low quality
basecalls (see [Post-alignment basecall quality filtering](#post-alignment-basecall-quality-filtering)).

_Optional outputs:_ 

Intermediate parsed and processed mutations associated
with individual reads can be optionally output by passing the
<kbd>--output-parsed-mutations</kbd> option
(see [file format](file_formats.md#parsed-mutations)).

Tables of mutation counts broken down by mutation classification
can be optionally output by passing the 
<kbd>--output-counted-mutations</kbd>
(see [file format](file_formats.md#mutation-counts)).

Tabular histograms of read lengths and per-read mutation counts
can be optionally output in the main ShapeMapper log by passing
the <kbd>--per-read-histograms</kbd>.

#### Rate calculation
At each position in a given RNA, mutation rate 
for each provided sample is calculated as the mutation count 
divided by the effective read depth at that position. In 
output tab-delimited `*_profile.txt` files, the relevant 
columns are `<sample>_mutations` (mutation counts), 
`<sample>_effective_depth` (effective read depths), and
`<sample>_rate` (mutation rate).

_Output plots:_
`*_profiles.pdf` output files include panels for read depths and
mutation rates. 

![read depths example](images/read_depths_TPP.png)

Read depths for each sample are plotted with solid lines,
and effective read depths are plotted with
light lines. 

![mutation rates example](images/mutation_rates_TPP.png)

Mutation rates plotted with
lines, and standard errors plotted as shaded areas. Mutation rate standard
errors are calculated as `sqrt(mutation rate) / sqrt(effective depth)`.

### Reactivity profile calculation and normalization

_Sample key:_ `M`: Modified, `U`: Untreated control, `D`: Denatured control

#### Reactivity
If three samples are provided, reactivity is calculated as
 
&emsp;&emsp; <tt>(rate<sub>M</sub> - rate<sub>U</sub>) / rate<sub>D</sub></tt>

If two samples are provided, reactivity is calculated as

&emsp;&emsp; <tt>rate<sub>M</sub> - rate<sub>U</sub></tt>

If only one sample is provided, reactivity is simply

&emsp;&emsp; <tt>rate<sub>M</sub></tt>

#### Standard error
If three samples are provided, reactivity stderr is calculated as

&emsp;&emsp; <tt>sqrt( stderr<sub>M</sub> / rate<sub>D</sub>)<sup>2</sup> +
               (stderr<sub>U</sub> / rate<sub>D</sub>)<sup>2</sup> +
               (stderr<sub>D</sub> × (rate<sub>M</sub> - rate<sub>U</sub>) / rate<sub>D</sub><sup>2</sup>)<sup>2</sup> )</tt>

where rate and rate stderr are calculated as [above](#calculation-of-mutation-rates).

If two samples are provided, reactivity stderr is calculated as

&emsp;&emsp; <tt>sqrt(stderr<sub>M</sub><sup>2</sup> + stderr<sub>U</sub><sup>2</sup>)</tt>

If only one sample is provided, reactivity stderr is simply

&emsp;&emsp; <tt>stderr<sub>M</sub></tt>

Reactivity standard errors are shown as error bars in `*_profile.pdf` plots.

#### Excluded positions
Lowercase sequence is excluded from reactivity profiles. Positions
with effective read depth in any sample below 
<kbd>--min-depth</kbd> 
(default=5000) are
excluded. If an untreated control sample is provided, positions with 
untreated mutation rate above 
<kbd>--max-bg</kbd> 
(default=0.05) are excluded.

#### Normalization
Over the set of all RNAs and nucleotide positions without 
masked (lowercase) sequence, high background, or low read depth, 
reactivities are normalized by dividing
by the mean reactivity of the top 10% of reactivities after 
reactivities above a threshold are excluded 
(see section 3.2.1 in [Low and Weeks, 2010](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2941709/)).
That threshold is selected from
the largest value out of

    [1.5 × interquartile range,
    90th percentile (if total seq length > 100) or 95th percentile (if total seq length < 100)].

By default, ShapeMapper normalizes all profiles together - 
that is, using the combined set of reactivities to compute 
a single normalization factor that is applied to all profiles.
This can be disabled by passing the <kbd>--indiv-norm</kbd> option.

_Example:_
![reactivity example](images/reactivity_TPP.png)

_Limitations:_

SHAPE-MaP measures relative probe reactivities, and normalization
is an attempt to put these relative values on a more meaningful
scale, closer to what might ideally be achieved with an absolute
measurement of nucleotide conformational state. However, this normalization 
relies on the assumption that the distribution of observed reactivities
in a particular experiment matches the distribution of reactivities
observed in experiments used to develop pseudo-free energy structure
modeling constraint parameters. In particular, normalization
 will produce systemically incorrect results for highly single-stranded RNAs or 
extremely highly base-paired RNAs.

Future work may attempt to fit relative reactivity distributions to empirical
models of the distributions of reactivities observed for paired and unpaired
nucleotides. This would allow more accurate 
structure modeling for highly structured or highly unstructured RNAs 
as well as providing an estimate of overall base pairing percentage.

## Quality control checks

ShapeMapper performs several basic quality-control checks for each 
reactivity profile. These are necessarily heuristic, since different
downstream analyses require different levels of data quality, and 
since RNAs of differing flexibility will show different overall
signal levels above background. In general, more sequencing read 
depth is always helpful, as are higher modification/mutation rates 
(to a point).

If ShapeMapper gives a red warning message about possible low-quality
reactivity profiles, read the log file to see which quality control
checks failed, and refer to this section for possible remedies.


#### Read depth check

ShapeMapper requires that at least 80% of non-masked nucleotides meet a minimum
sequencing depth of 5000 in all provided samples. Note that with
uneven sequencing coverage, some regions of an RNA may have 
higher-quality reactivity data than other regions. For example, 
coverage is often lower near transcript ends.

If read depths are low, check alignment stats in the logfile to see the 
amount of target sequence present in each sample. Better target enrichment or 
recovery and/or additional sequencing can often help.


#### Positive mutation rates above background check

If an untreated control sample is provided (highly recommended), ShapeMapper
requires that at least 50% of non-masked nucleotides with depths above
5000 have a higher mutation rate in the modified sample than in the 
untreated sample.


#### High background mutation rates check

If an untreated control sample is provided, ShapeMapper requires that no more
than 5% of non-masked nucleotide with depths above 5000 have an
untreated mutation rate above 0.05.

An unusual number of high-background nucleotides can result from the
presence of native modifications or, more often, a subpopulation of 
sequence variants or transcribed pseudogenes. There is no
built-in method to automatically determine the sequences of multiple
sequence variants in a mixture, but it can sometimes be helpful to
pass the <kbd>--render-mutations</kbd> option and visually inspect the PDF outputs 
from the untreated sample, looking for repeated instances of the same
mutation pattern from distinct reads. By default, these outputs are limited to
100 pages, controlled with the 
<kbd>--max-pages</kbd> parameter. The pages are wide enough
by default to accomodate a maximum paired read length of 800 nts, but this is
often very zoomed out; lower the 
<kbd>--max-paired-fragment-length</kbd> as needed for
visualization.


#### Number of highly reactive nucleotides check

ShapeMapper requires that at least 8% of non-masked nucleotides with depths 
above 5000 have a modified mutation rate above 0.006 after background
subtraction.

Possible causes for failure: 
- DNA contamination. Unusually low background mutation rates can 
  be a secondary indication that this is the problem (since 
  reverse transcription under MaP conditions usually generates
  some errors).
- Poor mixing of chemical reagents and RNA and/or poor reagent 
  diffusion (if modifying in cells), resulting in low 
  modification rates
- Expired reagents, resulting in low modification rates
- Poor reverse transcription conditions, resulting in low adduct 
  read-through
- Extremely highly structured RNA. A molecule that genuinely
  contains no flexible nucleotides is indistinguishable from a
  highly flexible one that was mistakenly unmodified (they will
  both have low mutation rates above background). In this case, 
  additional control experiments or complementary techniques may 
  be needed.
